helix-term/src/ui/picker.rs

```diff
@@ -346,22 +365,56 @@ impl<T> Picker<T> {
         self.cursor = 0;
     }
 
-    pub fn move_up(&mut self) {
+    /// Move the cursor by a number of lines, `amount`.
+    ///
+    /// Positive numbers move the cursor down, negative numbers move it up. The cursor cycles
+    /// through the entries.
+    pub fn move_by(&mut self, amount: isize) {
+        if self.matches.is_empty() {
+            return;
+        }
+
+        let cursor: isize = self.cursor.try_into().unwrap();
+        let len: isize = self.matches.len().try_into().unwrap();
+
+        self.cursor = cursor
+            .saturating_add(amount)
+            .rem_euclid(len)
+            .try_into()
+            .unwrap();
+    }
+
+    /// Move the cursor down by exactly one page. After the last page comes the first page.
+    pub fn page_up(&mut self) {
+        self.move_by(-(self.completion_height as isize));
+    }
+
+    /// Move the cursor up by exactly one page. After the first page comes the last page.
+    pub fn page_down(&mut self) {
+        self.move_by(self.completion_height as isize);
+    }
+
+    /// Move the cursor to the first entry
+    pub fn to_start(&mut self) {
         if self.matches.is_empty() {
             return;
         }
-        let len = self.matches.len();
-        let pos = ((self.cursor + len.saturating_sub(1)) % len) % len;
-        self.cursor = pos;
+        self.cursor = 0;
     }
 
-    pub fn move_down(&mut self) {
+    /// Move the cursor to the last entry
+    pub fn to_end(&mut self) {
         if self.matches.is_empty() {
             return;
         }
-        let len = self.matches.len();
-        let pos = (self.cursor + 1) % len;
-        self.cursor = pos;
+        self.cursor = self.matches.len() - 1;
```

Comment:
```suggestion
        self.cursor = self.matches.len().saturating_sub(1);
```

Otherwise this can panic if `len == 0`

Reply:
It can't panic because of the `if` branch in line 407, but I guess using `saturating_sub` is more idiomatic.

Reply:
It won't panic because `self.matches.is_empty()` will quick exit, but this might affect future developer since it relies on some explicit knowledge of the function, in case someone changed that part it will break.